File: ball_in_maze.py
class Solution(object):
    def hasPath(self, maze, start, destination):
        """
        :type maze: List[List[int]]
        :type start: List[int]
        :type destination: List[int]
        :rtype: bool
        """
        
        #bfs
        #O(mn) for both
        if not maze:
            return False
        q=deque()
        q.append(start)
        m=len(maze)
        n=len(maze[0])
        dirs=[[0,1],[1,0],[0,-1],[-1,0]]
        #mark visited>state change
        maze[start[0]][start[1]]=2
        while q:
            cur=q.popleft()
            #if matches with destination>return answer
            if cur[0]==destination[0] and cur[1]==destination[1]:
                return True
            #for each dir, try to go till extreme wall
            for dir in dirs:
                i=cur[0]+dir[0]
                j=cur[1]+dir[1]
                while (i>=0 and i<m and j>=0 and j<n) and maze[i][j]!=2:
                    i=i+dir[0]
                    j=j+dir[1]
                i=i-dir[0]
                j=j-dir[1]
                if maze[i][j]!=2:
                    q.append((i,j))
                    maze[i][j]=2
        return False
            

            
        
        

File: Ball-in-maze.java
//Solution without modifying input

//Time: O(m * n)
//Space: O(m * n)

class Solution {
  boolean ispath = false;
    public boolean hasPath(int[][] maze, int[] start, int[] destination) {
      int i=start[0]; int j=start[1];
      boolean[][] visited = new boolean[maze.length][maze[0].length];
  
      return dfs(maze,start[0],start[1],destination,visited); 
      
    }
  
    public boolean dfs(int[][] maze, int i,int j, int[] dest, boolean[][] visited){
     int[][] dirs = {{1,0},{-1,0},{0,1},{0,-1}};
      
      int m = maze.length;
      int n = maze[0].length;
      
      //Base cases
      if(visited[i][j] == true){
        ispath = false;
        return false;
      }
      
      if(i == dest[0] && j == dest[1]){
        ispath = true;
        return true;
      }
      
      visited[i][j] = true;
     
      for(int k=0;k<dirs.length;k++){
        int row = i;
        int col = j;
        // Keep rolling the ball till hit the wall
        while(row >=0 && row < m  && col >= 0 && col < n  && maze[row][col] != 1){
          // even if it is visited, its okay, can just roll over those cells
          row+= dirs[k][0];
          col+= dirs[k][1];
        }
        
        row -= dirs[k][0];
        col -= dirs[k][1];
   
        if(row >= m) row = m-1;
        if(row < 0) row = 0;
        if(col >= n) col = n-1;
        if(col < 0) col = 0;
        
        if(dfs(maze,row,col,dest,visited)){
          return true;
        }
        
        
      }
      
      return false;
    }
  
}

